cocotb/bfms.py

```diff
@@ -0,0 +1,206 @@
+        self.inst_name = inst_name
+        self.type_info = type_info
+        
+    def call_method(self, method_id, params):
+        self.type_info.export_info[method_id].T(
+            self.bfm, *params)
+
+class BfmMgr():
+    
+    m_inst = None
+    
+    def __init__(self):
+        self.bfm_l = []
+        self.bfm_type_info_m = {}
+        self.m_initialized = False
+    
+    def add_type_info(self, T, type_info):
+        self.bfm_type_info_m[T] = type_info
+    
+    @staticmethod
+    def get_bfms():
+        return BfmMgr.inst().bfm_l
+
+    @staticmethod    
+    def find_bfm(path_pattern):
+        inst = BfmMgr.inst()
+        bfm = None
+       
+        path_pattern_re = re.compile(path_pattern)
+        
+        for b in inst.bfm_l:
+            if path_pattern_re.match(b.bfm_info.inst_name):
+                bfm = b
+                break
+        
+        return bfm
+   
+    @staticmethod
+    def inst():
+        if BfmMgr.m_inst is None:
+            BfmMgr.m_inst = BfmMgr()
+            
+        return BfmMgr.m_inst
+   
+    def load_bfms(self):
+        '''
+        Obtain the list of BFMs from the native layer
+        '''
+        import simulator
+        n_bfms = simulator.bfm_get_count()
+        for i in range(n_bfms):
+            info = simulator.bfm_get_info(i)
+            instname = info[1]
+            clsname = info[2]
+            if clsname.find('.') == -1:
+                raise Exception("Incorrectly-formatted BFM class name {!r}".format(clsname))
+            pkgname = clsname[:clsname.rfind('.')]
+            clsleaf = clsname[clsname.rfind('.')+1:]
+          
+            try: 
+                pkg = importlib.import_module(pkgname)
+            except Exception as e:
+                raise Exception("Failed to import BFM package \"" + pkgname + "\"")
+       
+            if not hasattr(pkg, clsleaf):
+                raise Exception("Failed to find BFM class \"" + clsleaf + "\" in package \"" + pkgname + "\"") 
+
+            bfmcls = getattr(pkg, clsleaf)
+            
+            type_info = self.bfm_type_info_m[bfmcls]
+            
+            bfm = bfmcls()
+            bfm_info = BfmInfo(
+                bfm, 
+                len(self.bfm_l), 
+                instname,
+                type_info)
+            # Add 
+            setattr(bfm, "bfm_info", bfm_info)
+            
+            self.bfm_l.append(bfm)
+    
+    @staticmethod    
+    def init():
+        import simulator
+        inst = BfmMgr.inst()
+        if not inst.m_initialized:
+            simulator.bfm_set_call_method(BfmMgr.call)
+            BfmMgr.inst().load_bfms()
+            inst.m_initialized = True
+        
+    @staticmethod
+    def call(
+            bfm_id, 
+            method_id,
+            params):
+        inst = BfmMgr.inst()
+        bfm = inst.bfm_l[bfm_id]
+         
+        if not hasattr(bfm, "bfm_info"):
+            raise Exception("BFM object does not contain 'bfm_info' field")
```

Comment:
Maybe ``AttributeError`` would be better?

Reply:
Makes sense. Done